### ISSUE TITLE
Quick Dive into Task Tree Clustering

### ISSUE THREAD

**Issue by yoomlam** [department-of-veterans-affairs/caseflow#12665]
Initial dive into #12645. 
 
## Description

This ticket describes a small level of effort for clustering task trees to assess the utility of the results and feasibility of the approach. Immediate uses of the results include (1) automatic detection of invalid appeal states and (2) documentation of existent valid appeal states and state transitions from actual data, both of which will expedite diagnosis and remediation of user problems.
If results are positive, further work can be scoped out in other tickets. For more background, see #12645.

## Procedure

To minimize time spent on executing this ticket, each step is described in detail where possible to reduce research/exploration/investigation. Feedback to further expedite this effort is appreciated. Most of time will be spent on steps 4 and 6, exploring different distance metrics and clustering algorithms and examining the results.

1. Convert a task tree into a single-line JSON (for easy manipulation in any programming language). Ideally, we want to capture as much information as possible without running into PII issues. For this ticket, capturing some Appeal's fields and each tasks' `type`, `created_at`, `assigned_to_type`, and `status` is hopefully sufficient to determine the validity of an appeal. 
For example, given this appeal task tree:
```
Appeal 28 [assigned_to_type, status, created_at]
└── RootTask Organization, on_hold, 2019-11-05 20:35:00 UTC
    └── DistributionTask Organization, on_hold, 2019-11-05 20:35:00 UTC
        └── HearingTask Organization, on_hold, 2019-11-05 20:35:01 UTC
            ├── ScheduleHearingTask Organization, completed, 2019-11-05 20:35:01 UTC
            └── AssignHearingDispositionTask Organization, on_hold, 2019-11-05 20:35:01 UTC
                └── EvidenceSubmissionWindowTask Organization, assigned, 2019-11-05 20:35:01 UTC
```
it would be converted into something like this but in a single line:
```
{ docket_type: "hearing",
  appeal_id: 28, 
  created_at: "Tue, 05 Nov 2019 20:35:00 UTC +00:00", 
  status: "on_hold",
  tasks: [
    { type: "DistributionTask", assigned_to_type: "Organization", 
      status: "on_hold", created_at: "2019-11-05 20:35:00 UTC",
      tasks: [
        { type: "HearingTask", assigned_to_type: "Organization", 
          status: "on_hold", created_at: "2019-11-05 20:35:01 UTC",
          tasks: [
            { type: "ScheduleHearingTask", assigned_to_type: "Organization", 
              status: "completed", created_at: "2019-11-05 20:35:01 UTC"
            },
            { type: "AssignHearingDispositionTask", assigned_to_type: "Organization", 
              status: "on_hold", created_at: "2019-11-05 20:35:01 UTC",
              tasks: [
                { type: "EvidenceSubmissionWindowTask", assigned_to_type: "Organization", 
                  status: "assigned", created_at: "2019-11-05 20:35:01 UTC"
                }
            }
        }
    }
  ]
}
```

Note `appeal_id` is only needed for traceability in case we want to trace back to the actual appeal. If needed, the `appeal_id` can be left out. Traceability may still be possible if DB query results are consistently ordered and appeals are not deleted from the DB. If not, `created_at` may be sufficient for traceability. 
If needed, the `created_at` value can be obfuscated, but the relative ordering of task creation for an appeal should be retained for exploration in future tickets.

2. Dump all task trees (converted into a single-line JSON per appeal) into a file. This is the only time we query the DB. To avoid interrupting the live DB, (a) query against the replicated DB or (b) query at a low-use time period or (c) query in chunks.

3. Preprocess the data file for clustering. Typically, the data file would be converted into a CSV file where a column represents a distinct field from the task tree JSON; then various distance metrics can be tested for clustering (as typical). (We can even investigate representing the task tree with [RDF](http://ruby-rdf.github.io/) but since the task tree relationships are homogeneous, that's overkill.) 
Since we don't want to spend too much time extracting features and doing [feature selection](https://en.wikipedia.org/wiki/Feature_selection), let's ignore the `created_at` and `status` attributes, as well as the `appeal_id`. So for this initial trial, the JSON will look something like this (except it should be in one line):
```
{ docket_type: "hearing",
  tasks: [
    { type: "DistributionTask", assigned_to_type: "Organization",
      tasks: [
        { type: "HearingTask", assigned_to_type: "Organization",
          tasks: [
            { type: "ScheduleHearingTask", assigned_to_type: "Organization",
            },
            { type: "AssignHearingDispositionTask", assigned_to_type: "Organization",
              tasks: [
                { type: "EvidenceSubmissionWindowTask", assigned_to_type: "Organization",
                }
            }
        }
    }
  ]
}
```

As a result, many task trees will look exactly alike, thereby reducing the number of distinct data points (task trees) to process and assess. For detecting invalid task trees, keeping `status` may be useful -- something to try if time allows.

Note that traceability back to a specific task tree is preserved via the data point position/index in the list, as long as we keep the file created in the previous step. This index will be used to reference specific task trees for the remaining steps.

4. Define distance metric. Clustering requires a [distance metric](http://www.improvedoutcomes.com/docs/WebSiteDocs/Clustering/Clustering_Parameters/Distance_Metrics_Overview.htm) to compare data points. Since our data points are in JSON format, use one of the following:
- [deepdiff](https://github.com/seperman/deepdiff)
- [jsondiff](https://github.com/xlwings/jsondiff)
- [json-delta](https://json-delta.readthedocs.io/en/latest/json_diff.1.html)
- For consideration later, [graph isomorphism](https://gist.github.com/maetl/02a2b63346b27c7e449be09804e93245)

The output of these particular libraries is a list of differences between two JSON objects, so the output needs to be converted into a single number (a single [edit distance](https://en.wikipedia.org/wiki/Edit_distance)), such as number of additions + number of deletions + number of substitutions. Weighting of the terms can be incorporated, but ensure that the distance metric is symmetric/commutative.
If possible, task tree characteristics that are known to imply an invalid task tree should be weighted heavily. This will result in separating valid and invalid task trees, and/or clustering invalid task trees together.

5. Save distance matrix to a file. Since the data points are JSON strings, calculating the distance metric across a large number of data points could be computationally intensive, depending on how well the clustering library memoizes distance values. To avoid this cost, create and save a [distance matrix](https://en.wikipedia.org/wiki/Distance_matrix) using some of these resources:
- [pandas pairwise matrix how-to](https://www.drawingfromdata.com/making-a-pairwise-distance-matrix-with-pandas)
- [DIY code](http://code.activestate.com/recipes/498246-calculate-the-distance-matrix-for-n-dimensional-po/)
- [numpy distance matrix](http://people.duke.edu/~ccc14/sta-663-2016/03A_Numbers.html#Example:-Calculating-pairwise-distance-matrix-using-broadcasting-and-vectorization)
- [skbio.stats.distance.DistanceMatrix](http://scikit-bio.org/docs/0.2.0/generated/generated/skbio.stats.distance.DistanceMatrix.html)
- [scipy.spatial.distance.pdist](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.distance.pdist.html)
- [numpy.save](https://docs.scipy.org/doc/numpy/reference/generated/numpy.save.html): [note1](https://stackoverflow.com/questions/28439701/how-to-save-and-load-numpy-array-data-properly), [note2](https://towardsdatascience.com/why-you-should-start-using-npy-file-more-often-df2a13cc0161)

6. With the general distance matrix, explore different clustering algorithms. Determine if the clustering results are useful. Remember that clustering results can be changed with different distance metrics, so return to step 4 as desired.

- [intro](https://www.learndatasci.com/tutorials/k-means-clustering-algorithms-python-intro/)
- [k-means clustering](https://jakevdp.github.io/PythonDataScienceHandbook/05.11-k-means.html)
- [sklearn clustering methods](https://scikit-learn.org/stable/modules/clustering.html)

To determine utility, consider if the clustering results:
- distinguish different steps in the appeal process
- distinguish invalid vs. valid task trees. Since we don't have [labeled data](https://en.wikipedia.org/wiki/Labeled_data) for task tree validity, we may be able to assume that all clusters are valid and then decide that any new data point that doesn't fit in an existing cluster is invalid.
- (revisit #12645 to extract other uses)

We'll likely have to use different distance metrics and/or clustering algorithms for each listed utility above.

Other useful results:
- Data-driven state transition model: Use the `created_at` field of each task to capture how the state of each appeal changes over time (#12666). Do this for all valid appeals and union the state transitions to generate the state transition model. Appeals that do not follow this state transition model should be considered anomalous and the appeal state is potentially invalid -- see [outlier detection](https://hdbscan.readthedocs.io/en/latest/outlier_detection.html) or [paper](https://www.sciencedirect.com/science/article/pii/S1877050915005591).
- For each of the large clusters (those with the most data points), generate task tree diagram that is representative of that cluster. Include the diagram in documentation related to the appeal state. Use the `created_at` field to document how the appeal came into the state represented by the task tree. 

## Questions that will be answered (to some degree) upon completion

- [ ] How many clusters will there be to examine? There may be too many states/clusters to assess.
- [ ] Will there be useful clusters? Depends highly on the distance metric.
- [ ] How should we deal with old task trees with obsolete/defunct tasks and structure? They were valid in the past but are not relevant or compatible with current task trees.
- [ ] What issues for follow on work should be created?

**Comment by yoomlam** [comment]
If deeper analysis is desired, try examining changes to the location code in VACOLS mentioned in https://dsva.slack.com/archives/C3EAF3Q15/p1573667863033800:
```ruby
vacols_ids[0..3].each_with_index do |id, i|
  puts "vacols_id=#{id}"
  pp VACOLS::Priorloc.where(lockey: id).order(:locdout).pluck(:locdout, :locstto, :locstrcv)
end
```
Example location history:
```
[[Fri, 18 Mar 2016, "77", "IMBUTLER"],
 [Mon, 11 Feb 2019, "78", "VACOLEEC4"],
 [Wed, 13 Feb 2019, "01", "VACOLEEC4"],
 [Wed, 13 Feb 2019, "57", "DSUSER"],
 [Fri, 15 Feb 2019, "CASEFLOW", "SAUNDR3"],
 [Tue, 12 Mar 2019, "CASEFLOW", "SAUNDR3"],
 [Wed, 03 Apr 2019, "CASEFLOW", "SAUNDR3"],
 [Wed, 03 Apr 2019, "81", "DAMES"],
 [Thu, 04 Apr 2019, "DAMES", "DAMES"],
 [Thu, 04 Apr 2019, "VACOFOWLEJ", "VACOFOWLEJ"],
 [Mon, 08 Apr 2019, "12", "TOWENS"],
 [Wed, 10 Apr 2019, "57", "DSUSER"],
 [Mon, 10 Jun 2019, "CASEFLOW", "VACOCARTEL"],
 [Wed, 12 Jun 2019, "CASEFLOW", "VACOCARTEL"],
 [Fri, 19 Jul 2019, "CASEFLOW", "VACOGRAYB1"],
 [Fri, 23 Aug 2019, "CASEFLOW", "VACOGRAYB1"],
 [Fri, 23 Aug 2019, "81", "VACOGRAYB1"],
 [Fri, 23 Aug 2019, "81", "DWARE"],
 [Wed, 28 Aug 2019, "82", "VACOARTHUA"],
 [Fri, 30 Aug 2019, "83", "JWILLS"],
 [Wed, 04 Sep 2019, "JWILLS", "JWILLS"],
 [Mon, 09 Sep 2019, "BRIDEOUT", "BRIDEOUT"],
 [Fri, 20 Sep 2019, "DAMES", "DAMES"],
 [Tue, 24 Sep 2019, "JWILLS", "JWILLS"],
 [Tue, 01 Oct 2019, "4E", "VACOCURTIS"],
 [Tue, 01 Oct 2019, "97", "MWILLIAMS"],
 [Wed, 02 Oct 2019, "50", nil]]
```

**Comment by lomky** [comment]
_[What is this graph?](https://github.com/department-of-veterans-affairs/caseflow/issues/11533)_
```
1 | 
2 | 
3 | ||||||
5 | |||||||
8 | |||||
```

Putting at a 5. We feel like this is the minimal worthwhile timebox. Debating splitting this up into smaller workable chunks?
